sklearn/metrics/_pairwise_distances_reduction/_pairwise_distances.pyx.tp

```diff
@@ -0,0 +1,373 @@
+
+        No instance should directly be created outside of this class method.
+        """
+        if (
+            metric in ("euclidean", "l2", "sqeuclidean")
+            and not issparse(X)
+            and not issparse(Y)
+        ):
+            # Specialized implementation with improved arithmetic intensity
+            # and vector instructions (SIMD) by processing several vectors
+            # at time to leverage a call to the BLAS GEMM routine as explained
+            # in more details in the docstring.
+            use_squared_distances = metric == "sqeuclidean"
+            pdr = EuclideanPairwiseDistances{{name_suffix}}(
+                X=X, Y=Y,
+                use_squared_distances=use_squared_distances,
+                chunk_size=chunk_size,
+                metric_kwargs=metric_kwargs,
+                strategy=strategy,
+            )
+        else:
+            # Precompute data-derived distance metric parameters
+            metric_kwargs = {} if metric_kwargs is None else metric_kwargs
+
+            params = _precompute_metric_params(
+                X,
+                Y,
+                metric=metric,
+                **metric_kwargs,
+            )
+            metric_kwargs.update(**params)
+
+            # Fall back on a generic implementation that handles most scipy
+            # metrics by computing the distances between 2 vectors at a time.
+            pdr = PairwiseDistances{{name_suffix}}(
+                datasets_pair=DatasetsPair{{name_suffix}}.get_for(X, Y, metric, metric_kwargs),
+                chunk_size=chunk_size,
+                strategy=strategy,
+            )
+
+        # Limit the number of threads in second level of nested parallelism for BLAS
+        # to avoid threads over-subscription (in GEMM for instance).
+        with threadpool_limits(limits=1, user_api="blas"):
+            if pdr.execute_in_parallel_on_Y:
+                pdr._parallel_on_Y()
+            else:
+                pdr._parallel_on_X()
+
+        return pdr._finalize_results()
+
+
+    def __init__(
+        self,
+        DatasetsPair{{name_suffix}} datasets_pair,
+        chunk_size=None,
+        strategy=None,
+        sort_results=False,
+    ):
+        super().__init__(
+            datasets_pair=datasets_pair,
+            chunk_size=chunk_size,
+            strategy=strategy,
+        )
+
+        # Distance matrix which will be complete and returned to the caller.
+        self.pairwise_distances_matrix = np.empty(
+            (self.n_samples_X, self.n_samples_Y), dtype=DTYPE,
+        )
+
+    def _finalize_results(self):
+        # If X is Y, then catastrophic cancellation might
+        # have occurred for computations of term on the diagonal
+        # which must be null. We enforce nullity of those term
+        # by zeroing the diagonal.
+        distance_matrix = np.asarray(self.pairwise_distances_matrix)
+        if self.datasets_pair.X_is_Y:
+            np.fill_diagonal(distance_matrix, 0.)
```

Comment:
Potentially, we might want (remove this and) add a guard directly after computing distances to prevent negative zeros and NaN [as done in the Euclidean Variant](https://github.com/jjerphan/scikit-learn/blob/05b5ae7774425c4e07004fd0967a85c71c124d4e/sklearn/metrics/_pairwise_distances_reduction/_pairwise_distances.pyx.tp#L357-L361).

If removing this part, then `DatasetsPair.X_is_Y` must also be removed.

What do you think?

Reply:
`_euclidean_distances` does prevent negative values. I think we should keep that behavior

Reply:
We should generalize that to all metrics (if they are true metrics). Since this not used only in `pairwise_distances` I guess we have to do that here to avoid code duplication

Reply:
What I meant is that we should also propagate the [guard against negative zeros and NaNs in `EuclideanPairwiseDistances`](https://github.com/jjerphan/scikit-learn/blob/05b5ae7774425c4e07004fd0967a85c71c124d4e/sklearn/metrics/_pairwise_distances_reduction/_pairwise_distances.pyx.tp#L357-L361) to other `Euclidean` variant because they are not  all protected against NaNs and negative zeros.